extensions/src/main/java/org/pentaho/platform/plugin/services/security/userrole/oauth/PentahoOAuthOktaHandler.java

```diff
@@ -0,0 +1,173 @@
+/*! ******************************************************************************
+ *
+ * Pentaho
+ *
+ * Copyright (C) 2024 by Hitachi Vantara, LLC : http://www.pentaho.com
+ *
+ * Use of this software is governed by the Business Source License included
+ * in the LICENSE.TXT file.
+ *
+ * Change Date: 2029-07-20
+ ******************************************************************************/
+
+package org.pentaho.platform.plugin.services.security.userrole.oauth;
+
+import org.apache.commons.lang.StringUtils;
+import org.pentaho.platform.api.engine.security.IAuthenticationRoleMapper;
+import org.pentaho.platform.api.engine.security.userroledao.IUserRoleDao;
+import org.pentaho.platform.api.mt.ITenant;
+import org.pentaho.platform.security.userroledao.PentahoOAuthUser;
+import org.pentaho.platform.util.oauth.PentahoOAuthProperties;
+import org.pentaho.platform.util.oauth.PentahoOAuthUtility;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class PentahoOAuthOktaHandler implements IPentahoOAuthHandler {
+
+  IUserRoleDao userRoleDao;
+
+  RestTemplate restTemplate;
+
+  IAuthenticationRoleMapper oauthRoleMapper;
+
+  PentahoOAuthProperties pentahoOAuthProperties;
+
+  Map<String, String> registrationToClientCredentialsToken = new HashMap<>();
+
+  public PentahoOAuthOktaHandler( IUserRoleDao userRoleDao,
+                                  RestTemplate restTemplate,
+                                  IAuthenticationRoleMapper oauthRoleMapper,
+                                  PentahoOAuthProperties pentahoOAuthProperties ) {
+    this.userRoleDao = userRoleDao;
+    this.restTemplate = restTemplate;
+    this.oauthRoleMapper = oauthRoleMapper;
+    this.pentahoOAuthProperties = pentahoOAuthProperties;
+  }
+
+  public String getClientCredentialsToken( String registrationId, boolean renewToken ) {
+    String clientCredentialsToken = registrationToClientCredentialsToken.get( registrationId );
+
+    if ( renewToken || StringUtils.isBlank( clientCredentialsToken ) ) {
+      HttpHeaders headers = new HttpHeaders();
+      headers.setContentType( MediaType.APPLICATION_FORM_URLENCODED );
+      headers.setBasicAuth( pentahoOAuthProperties.getValue( registrationId + ".client-id" ),
+              pentahoOAuthProperties.getValue( registrationId + ".client-secret" ) );
+
+      MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
+      map.add( "grant_type", "client_credentials" );
```

Comment:
can you makes thes map keys java constants, there seems to be overlap between Okta and Azure handler code.

Reply:
addressed